src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.RemoveUnnecessaryParentheses
+{
+    internal abstract class AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer<
+        TLanguageKindEnum,
+        TParenthesizedExpressionSyntax>
+        : AbstractParenthesesDiagnosticAnalyzer
+        where TLanguageKindEnum : struct
+        where TParenthesizedExpressionSyntax : SyntaxNode
+    {
+        protected AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveUnnecessaryParenthesesDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Remove_unnecessary_parentheses), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Parentheses_can_be_removed), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public sealed override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
```

Comment:
InitializeWorker is not a super great name, it sounds liek there will be some threading/background tasks going on

Reply:
This is the name we have for this hierarchy.  i do not want to change it.  "Worker" is the general naming pattern we've chosen in many places for "abstract helper called from parent to do work specific to each subclass".